bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryClientFactory.java

```diff
@@ -86,4 +168,40 @@ public void setProxyConfig(InstantiatingGrpcChannelProvider.Builder transportBui
               proxyConfig.getProxyPassword()));
     }
   }
+
+  private void shutdownBigQueryReadClient(BigQueryReadClient bigQueryReadClient) {
+    if (bigQueryReadClient != null && !bigQueryReadClient.isShutdown()) {
+      bigQueryReadClient.shutdown();
+    }
+  }
+
+  private void shutdownBigQueryWriteClient(BigQueryWriteClient bigQueryWriteClient) {
+    if (bigQueryWriteClient != null && !bigQueryWriteClient.isShutdown()) {
+      bigQueryWriteClient.shutdown();
+    }
+  }
+
+  private boolean areCredentialsEqual(Credentials credentials1, Credentials credentials2) {
+    if (!(credentials1 instanceof ExternalAccountCredentials)
+        && !(credentials2 instanceof ExternalAccountCredentials)) {
+      return Objects.equal(credentials1, credentials2);
+    }
+
+    try {
+      ByteArrayOutputStream byteArrayOutputStream1 = new ByteArrayOutputStream();
+      ObjectOutputStream objectOutputStream1 = new ObjectOutputStream(byteArrayOutputStream1);
+      objectOutputStream1.writeObject(credentials1);
+      objectOutputStream1.close();
+
+      ByteArrayOutputStream byteArrayOutputStream2 = new ByteArrayOutputStream();
```

Comment:
Duplicated code, lines 191-194, 196-199 can be extracted to another method returning `byte[]` (which can also be used by `hashCode()`, see above)